source/geod24/concurrency.d

```diff
@@ -1446,3 +1502,665 @@ private:
+        }
+
+        /***********************************************************************
+
+            Terminate a `ChannelQueueEntry` by waking up the blocked Fiber
+            and reporting the failure
+
+            This is called on all the `ChannelQueueEntry` instances still in
+            the wait queues when Channel is closed
+
+        ***********************************************************************/
+
+        void terminate () nothrow
+        {
+            this.success = false;
+            this.sem.notify();
+        }
+    }
+
+    /***********************************************************************
+
+        Create and enqueue a `ChannelQueueEntry` to the given entryq
+
+        Assumes that the Channel is locked, unlocks it before returning
+
+        Param:
+            entryq = Queue to append the new ChannelQueueEntry
+            pVal =  Pointer to the message buffer
+            sel_state = SelectState object to associate with the
+                        newly created ChannelQueueEntry
+            sel_id = id of the select call creating the new ChannelQueueEntry
+        Return:
+            newly created ChannelQueueEntry
+
+    ***********************************************************************/
+
+    private ChannelQueueEntry enqueueEntry (ref DList!ChannelQueueEntry entryq, T* pVal) nothrow
+    {
+        assert(pVal !is null);
+
+        ChannelQueueEntry q_ent = new ChannelQueueEntry(pVal);
+        entryq.insert(q_ent);
+
+        return q_ent;
+    }
+
+    /***********************************************************************
+
+        Dequeue a `ChannelQueueEntry` from the given `entryq`
+
+        Walks the `entryq` until it finds a suitable entry or the queue
+        empties. If `dequeueEntry` is called from a select, it tries to
+        consume the `caller_sel` if the `peer_sel` is not currently consumed.
+        If it fails to consume the `caller_sel`, returns with a failure.
+
+        If selected queue entry is part of a select operation, it is also
+        consumed. If it consumes `caller_sel` but `peer_sel` was already
+        consumed whole select operation would fail and caller would need to try
+        again. This should be a rare case, where the `peer_sel` gets consumed by
+        someone else between the first if check which verifies that it is not
+        consumed and the point we actually try to consume it.
+
+        Param:
+            entryq = Queue to append the new ChannelQueueEntry
+            caller_sel = SelectState instace of the select call being executed
+            caller_sel_id = id of the select call being executed
+
+        Return:
+            a valid ChannelQueueEntry or null
+
+    ***********************************************************************/
+
+    private ChannelQueueEntry dequeueEntry (ref DList!ChannelQueueEntry entryq) nothrow
+    {
+        while (!entryq.empty())
```

Comment:
This should be an `if`.

Reply:
Yep, that must have left over while manually I split the code into commits :(